packages/@lwc/template-compiler/src/codegen/optimize.ts

```diff
@@ -0,0 +1,111 @@
+ *   };
+ * }
+ * ```
+ *
+ * Example output:
+ *
+ * ```
+ * const stc0 = { static: true };
+ * const stc1 = { really: { static: ['yep', 'totally', 'static' ] } };
+ * function tmpl() {
+ *   return {
+ *     foo: dynamic(),
+ *     bar: stc0,
+ *     baz: stc1
+ *   };
+ * }
+ * ```
+ */
+export function optimizeStaticExpressions(
+    templateFn: t.FunctionDeclaration
+): Array<t.FunctionDeclaration | t.VariableDeclaration> {
+    const result: Array<t.FunctionDeclaration | t.VariableDeclaration> = [];
+    const keysToVariableNames = new Map();
+
+    let identifierCount = 0;
+
+    // Return true if this node is an object/array that is fully static
+    function isStaticObjectOrArray(node: t.BaseNode): boolean {
+        if (t.isObjectExpression(node)) {
+            return node.properties.every((prop) => {
+                return (
+                    t.isProperty(prop) &&
+                    !prop.computed &&
+                    !prop.method &&
+                    !prop.shorthand &&
+                    (t.isLiteral(prop.value) || isStaticObjectOrArray(prop.value))
+                );
+            });
+        } else if (t.isArrayExpression(node)) {
+            return node.elements.every((element) => {
+                return (
+                    element !== null &&
+                    t.isExpression(element) &&
+                    (t.isLiteral(element) || isStaticObjectOrArray(element))
+                );
+            });
+        }
+        return false;
+    }
+
+    function extractStaticVariable(node: t.BaseNode): t.BaseNode {
+        // This key generation can probably be improved using a hash code, but stringification is
+        // simplest for finding a unique identifier for an object/array expression
+        const key = astring.generate(node);
+
+        // Check for duplicates to avoid re-declaring the same object/array multiple times
+        // Especially for the empty array (`[]`), which is very common in templates
+        if (!keysToVariableNames.has(key)) {
+            const variableName = `stc${identifierCount++}`;
+            // e.g. `const stc0 = { /* original object */ };
+            const declaration = t.variableDeclaration('const', [
+                t.variableDeclarator(
+                    t.identifier(variableName),
+                    node as t.ArrayExpression | t.ObjectExpression
+                ),
+            ]);
+            result.push(declaration);
+            keysToVariableNames.set(key, variableName);
+        }
+
+        return t.identifier(keysToVariableNames.get(key));
+    }
+
+    walk(templateFn, {
+        enter(node) {
+            // For deeply-nested static object, we only want to extract the top-level node
+            if (isStaticObjectOrArray(node)) {
+                const newNode = extractStaticVariable(node);
+                this.replace(newNode);
+                this.skip();
+            }
+        },
+    });
```

Comment:
This approach looks reasonable. Do you think it is feasible to do this as we are building the AST and keep this in the `CodeGen` object to avoid the extra pass on the AST?

I would love in the future to simply get rid of the ESTree and directly produce a JavaScript string directly. The direct code generation would make it impossible to do this extra pass.

Reply:
It would be possible, but I figured that this approach would make it easier to remove this later if it causes problems, or to augment it. I see it as an "optimization" or transpilation step where we can potentially do many things – hoisting, deduplicating, etc.

Reply:
Sounds good to me. It's something we can improve later on if needed.

Could you check the performance benchmarks on the LWC platform compiler to make sure the potential performance hit is under control? I am interested in the added performance overhead for large HTML templates.